File: 2022/04/4.1.js
const fs = require('fs');

const solve = (input) => {
    const res = parseRows(input).reduce(findFullOverlaps, [])
    return res.length
}

const findFullOverlaps = (acc = [], [[x1,x2],[y1,y2]]) => {
    if ((x1 >= y1 && x2 <= y2) || (y1 >= x1 && y2 <= x2)) {
        acc.push([[x1,x2],[y1,y2]])
    }
    return acc
}

const  parseRows = (input) => input.split("\n").map(x => x.split(",").map(y => y.split("-").map(parseFloat)))

console.log(solve(fs.readFileSync(__dirname + '/4test.txt', 'utf8')) == 2);
console.log(solve(fs.readFileSync(__dirname + '/4.txt', 'utf8')));

File: 2020/04/4.1.rb
require_relative 'helpers.rb'

module DayFourPartOne
  module_function

  def solve(file)
    inputs = prepare_inputs(file)
    find_valid_passports(inputs)
  end

  def find_valid_passports(inputs)
    inputs.reduce(0) do |result, input|
      result += 1 if (%i[
        byr
        iyr
        eyr
        hgt
        hcl
        ecl
        pid
      ] - input.keys).empty?

      result
    end
  end

  def prepare_inputs(file)
    Helpers.read_file(file).split("\n\n")
           .map do |inputs|
      inputs.split(/[\n ]/)
            .each_with_object({}) do |input, result|
        split_input = input.split(':')
        result[split_input[0].to_sym] = split_input[1]
      end
    end
  end
end

# puts DayFourPartOne.solve(__dir__ + '/4test.txt')
puts DayFourPartOne.solve(__dir__ + '/4.txt')
